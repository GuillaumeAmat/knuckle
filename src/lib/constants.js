const MERGE_STRATEGIES = {
  DEEP: 'deep',
  SPREAD: 'spread',
  REPLACE: 'replace',
};

const DEFAULT_MERGE_STRATEGY = MERGE_STRATEGIES.DEEP;

module.exports = {
  MERGE_STRATEGIES,
  DEFAULT_MERGE_STRATEGY,
};
