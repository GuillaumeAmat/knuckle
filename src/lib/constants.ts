export const MERGE_STRATEGIES = {
  DEEP: 'deep',
  SPREAD: 'spread',
  REPLACE: 'replace',
};

export const DEFAULT_MERGE_STRATEGY = MERGE_STRATEGIES.DEEP;

export type Json = string | number | boolean | null | JsonObject | JsonArray;
export type JsonArray = Json[];
export type JsonObject = { [property: string]: Json };

export type ConfigurationGenerator = (
  mergeStrategy: string,
) => Array<{
  filename: string;
  build(configuredTools: string[]): string | Json | Array<any>;
  format(config: Json): Json | Array<any>;
}>;

export type DependenciesGetter = (configuredTools: string[]) => string[];
