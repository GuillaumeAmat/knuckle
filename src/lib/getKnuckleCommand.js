const { isKnuckleInKnuckle } = require('./isKnuckleInKnuckle');

/**
 * Returns the right command command to call Knuckle,
 * depending if the current process is in a Knuckle repository clone
 * or in another project.
 */
function getKnuckleCommand() {
  return isKnuckleInKnuckle() ? 'yarn knuckle' : 'npx knuckle';
}

module.exports = { getKnuckleCommand };
