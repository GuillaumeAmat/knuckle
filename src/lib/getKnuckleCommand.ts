import { isKnuckleInKnuckle } from './isKnuckleInKnuckle';

/**
 * Returns the right command command to call Knuckle,
 * depending if the current process is in a Knuckle repository clone
 * or in another project.
 */
export function getKnuckleCommand() {
  return isKnuckleInKnuckle() ? 'yarn knuckle' : 'npx knuckle';
}
