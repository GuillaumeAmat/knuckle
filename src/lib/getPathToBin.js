const path = require('path');

/**
 * Given a module name, that function returns its binary path
 * Eg:
 * - getPathToBin('prettier')
 * - getPathToBin('@commitlint/cli', 'commitlint')
 *
 * @param {string} moduleName
 * @param {string} commandName
 */
function getPathToBin(moduleName, commandName) {
  const modPkgPath = require.resolve(`${moduleName}/package.json`);
  const modPkgDir = path.dirname(modPkgPath);
  const { bin } = require(modPkgPath);
  const binPath = typeof bin === 'string' ? bin : bin[commandName || moduleName];

  return path.join(modPkgDir, binPath);
}

module.exports = { getPathToBin };
