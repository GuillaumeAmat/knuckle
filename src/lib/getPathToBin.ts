import path from 'path';

/**
 * Given a module name, that function returns its binary path
 * Eg:
 * - getPathToBin('prettier')
 * - getPathToBin('@commitlint/cli', 'commitlint')
 */
export function getPathToBin(moduleName: string, commandName?: string) {
  const modPkgPath = require.resolve(`${moduleName}/package.json`);
  const modPkgDir = path.dirname(modPkgPath);
  const { bin } = require(modPkgPath);
  const binPath = typeof bin === 'string' ? bin : bin[commandName || moduleName];

  return path.join(modPkgDir, binPath);
}
