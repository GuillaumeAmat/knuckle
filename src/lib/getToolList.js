const fs = require('fs');
const path = require('path');

/**
 * Returns the complete list of Knuckle supported tools
 */
function getToolList() {
  return fs.readdirSync(path.join(__dirname, '../tools'));
}

module.exports = { getToolList };
