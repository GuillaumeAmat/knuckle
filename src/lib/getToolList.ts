import fs from 'fs';
import path from 'path';

/**
 * Returns the complete list of Knuckle supported tools
 */
export function getToolList() {
  return fs.readdirSync(path.join(__dirname, '../tools'));
}
