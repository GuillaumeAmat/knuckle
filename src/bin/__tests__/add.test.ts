import { binScriptPath } from '../../helpers/tests/binScriptPath';
import * as cmd from '../../helpers/tests/cmd';
import { getKnuckleConfig } from '../../helpers/tests/getKnuckleConfig';
import { setupPristineTestFolder } from '../../helpers/tests/setupPristineTestFolder';
import { getToolList } from '../../lib/getToolList';

describe('ADD', () => {
  setupPristineTestFolder();

  it('should not generate a config file', async done => {
    await cmd.run(
      'node',
      [binScriptPath, 'add'],

      // Validate no previous selection
      [cmd.ENTER],
    );

    const configFile = getKnuckleConfig();

    expect(configFile).toBeNull();

    done();
  });

  it('should add tools interactively', async done => {
    await cmd.run(
      'node',
      [binScriptPath, 'add'],

      // Selects the 3 first tools and validate
      [cmd.SPACE, cmd.DOWN, cmd.SPACE, cmd.DOWN, cmd.SPACE, cmd.DOWN, cmd.ENTER],
    );

    const config = getKnuckleConfig()?.config;

    expect(config.tools.length).toBe(3);

    done();
  });

  it('should add tools with one command', async done => {
    await cmd.run('node', [binScriptPath, 'add', 'prettier', 'eslint', 'lint-staged', 'husky']);

    const config = getKnuckleConfig()?.config;

    expect(config.tools.length).toBe(4);

    done();
  });

  it('should not duplicate tools', async done => {
    // First call
    await cmd.run('node', [binScriptPath, 'add', 'prettier', 'eslint', 'lint-staged', 'husky']);

    // Second consecutive call with the same tools
    await cmd.run('node', [binScriptPath, 'add', 'prettier', 'eslint', 'lint-staged', 'husky']);
    const config = getKnuckleConfig()?.config;
    expect(config.tools.length).toBe(4);

    // Third call with duplicate tools and new ones
    await cmd.run('node', [binScriptPath, 'add', 'prettier', 'commitlint']);
    const newConfig = getKnuckleConfig()?.config;
    expect(newConfig.tools.length).toBe(5);

    done();
  });

  it('should tell if all the tools are handled', async done => {
    await cmd.run('node', [binScriptPath, 'add', ...getToolList()]);

    // With a specified tool...
    let response = await cmd.run('node', [binScriptPath, 'add', 'prettier']);

    expect(response).toMatchSnapshot();

    // ... And without
    response = await cmd.run('node', [binScriptPath, 'add']);

    expect(response).toMatchSnapshot();

    done();
  });

  it('should complain if we ask for non-supported tool', async done => {
    try {
      await cmd.run('node', [binScriptPath, 'add', 'non-supported-tool']);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }

    expect.hasAssertions();

    done();
  });
});
