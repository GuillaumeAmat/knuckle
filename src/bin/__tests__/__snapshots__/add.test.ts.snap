// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ADD should complain if we ask for non-supported tool 1`] = `
"
"
`;
